web/toolsupport.go

```diff
@@ -0,0 +1,104 @@
+package web
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/boltdb/bolt"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/zenodb/encoding"
+)
+
+// PermalinkInfo provides info about a stored permalink
+type PermalinkInfo struct {
+	Permalink string
+	SQL       string
+	Timestamp time.Time
+	NumRows   int
+}
+
+type permalinkInfo struct {
+	Permalink string
+	SQL       string
+	TS        int64
+	Rows      []*json.RawMessage
+}
+
+type PermalinkInfoRow struct {
+	TS int64
+}
+
+type byTimestamp []*PermalinkInfo
+
+func (a byTimestamp) Len() int           { return len(a) }
+func (a byTimestamp) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byTimestamp) Less(i, j int) bool { return a[i].Timestamp.After(a[j].Timestamp) }
+
+// ListPermalinks returns a list of known permalinks and associated info
+func ListPermalinks(cacheFile string) ([]*PermalinkInfo, error) {
+	c, err := newCacheFromFile(cacheFile, 24*time.Hour, &bolt.Options{
+		ReadOnly: true,
+		Timeout:  10 * time.Second,
+	})
+	if err != nil {
+		return nil, errors.New("Unable to open cache at %v: %v", cacheFile, err)
+	}
+	permalinks, err := c.listPermalinks()
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(byTimestamp(permalinks))
+	return permalinks, nil
+}
+
+func (c *cache) listPermalinks() ([]*PermalinkInfo, error) {
+	var permalinks []*PermalinkInfo
+	err := c.db.View(func(tx *bolt.Tx) error {
```

Comment:
Nit: If `listPermalinks` is conceptually considered a feature of the `cache`, the code should be in `cache.go`; if not, maybe make it a reguar function, i.e., `listPermalinks(c *cache)`. Note this is the only case `c` is referenced.

Reply:
The function needs to be public and cache is not. Conceptually, it only exists to support zenotool, hence it's in `toolsupport.go`.

Reply:
Okay